builtin/credential/approle/path_role.go

```diff
@@ -0,0 +1,1544 @@
+				logical.DeleteOperation: b.pathRoleSecretIDAccessorDelete,
+			},
+			HelpSynopsis:    strings.TrimSpace(roleHelp["role-secret-id-accessor"][0]),
+			HelpDescription: strings.TrimSpace(roleHelp["role-secret-id-accessor"][1]),
+		},
+		&framework.Path{
+			Pattern: "role/" + framework.GenericNameRegex("role_name") + "/custom-secret-id$",
+			Fields: map[string]*framework.FieldSchema{
+				"role_name": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: "Name of the role.",
+				},
+				"secret_id": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: "SecretID to be attached to the role.",
+				},
+				"metadata": &framework.FieldSchema{
+					Type: framework.TypeString,
+					Description: `Metadata to be tied to the SecretID. This should be a JSON
+formatted string containing metadata in key value pairs.`,
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: b.pathRoleCustomSecretIDUpdate,
+			},
+			HelpSynopsis:    strings.TrimSpace(roleHelp["role-custom-secret-id"][0]),
+			HelpDescription: strings.TrimSpace(roleHelp["role-custom-secret-id"][1]),
+		},
+	}
+}
+
+// pathRoleExistenceCheck returns whether the role with the given name exists or not.
+func (b *backend) pathRoleExistenceCheck(req *logical.Request, data *framework.FieldData) (bool, error) {
+	role, err := b.roleEntry(req.Storage, data.Get("role_name").(string))
+	if err != nil {
+		return false, err
+	}
+	return role != nil, nil
+}
+
+// pathRoleList is used to list all the Roles registered with the backend.
+func (b *backend) pathRoleList(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	b.roleLock.RLock()
+	defer b.roleLock.RUnlock()
+	roles, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(roles), nil
+}
+
+// pathRoleSecretIDList is used to list all the 'secret_id_accessor's issued against the role.
+func (b *backend) pathRoleSecretIDList(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	// Get the role entry
+	role, err := b.roleEntry(req.Storage, strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("role %s does not exist", roleName)), nil
+	}
+
+	// If the argument to secretIDLock does not start with 2 hex
+	// chars, a generic lock is returned. So, passing empty string
+	// to get the "custom" lock that could be used for listing.
+	lock := b.secretIDLock("")
+	lock.RLock()
+	defer lock.RUnlock()
+
+	// Listing works one level at a time. Get the first level of data
+	// which could then be used to get the actual SecretID storage entries.
+	secretIDHMACs, err := req.Storage.List(fmt.Sprintf("secret_id/%s/", b.salt.SaltID(role.RoleID)))
```

Comment:
This indicates that the path uses the RoleID -- secret_id/<salted role id>. What happens if they give the same role ID to two roles? The salt is per-backend instead of per-role, so AFAICT you'll end up listing secret IDs for more than one role.